classes/observers/events/BasePublicationEvent.inc.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @file classes/observers/events/BasePublicationEvent.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class BasePublicationEvent
+ * @ingroup core
+ *
+ * @brief Base event for publications being published or unpublished
+ */
+
+namespace PKP\observers\events;
+
+use Illuminate\Foundation\Events\Dispatchable;
+
+use PKP\publication\PKPPublication;
+use PKP\submission\PKPSubmission;
+
+class BasePublicationEvent
```

Comment:
Is `BasePublicationEvent` used as an actual event somewhere? If not, I'd recommend moving it away from the `events` folder and making abstract

Reply:
It's extended at `PublishedEvent` and `UnpublishedEvent`.

Reply:
Yeah, I lean towards keeping in `observers/events` only those that are used in the system and, in this case, moving `BasePublicationEvent` from there with declaring this class as abstract.

Reply:
@henriqueramos and @Vitaliy-1, is this resolved?

Reply:
@henriqueramos, sorry for not being clear during our meeting.
To avoid code duplication in the case of Event classes, I recommend using traits rather than extending a base class. E.g., something like:
```php
namespace PKP\observers; // outside events or listeners dir

trait Publicationable
{
  public PKPPublication $newPublication;
  public PKPPublication $publication;
  public PKPSubmission $submission;

  public function __construct(PKPPublication $newPublication, PKPPublication $publication, PKPSubmission $submission)
  {
     ...
  }
}
```
and then we can use:
```php
class PublishedEvent
{
	use Dispatchable;
	use Publicationable;
}
```
and for the listener part:
```php
class SubmissionUpdated
{
   /**
   * @param \PKP\observers\events\PublishedEvent $event
   */
  public function handle(PublishedEvent $event)
  {
     dispatch(new UpdateSubmissionSearchJob($event->submission->getId()));
   }
}
```
or if it listening to several events, use subscriber
```php
class SubmissionUpdated
{
   public function subscribe(Dispatcher $events): void
  {
     $events->listen(
         PublishedEvent::class,
         self::class . '@handlePublished'
     );

     $events->listen(
         UnpublishedEvent::class,
         self::class . '@handleUnpublished'
     );
   }

  public function handlePublished(PublishedEvent $event)...

  public function handleUnpublished(UnpublishedEvent $event)... 
}
```
This makes it clear which event listener actually listens.
Let me know if you'd need some additional info here

Reply:
Thank you for the idea of using Trait on this @Vitaliy-1! :tada: